## **分析目标**

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。  

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## **简介**

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚几年礼品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。  

数据每列的含义如下：  
 - InvoIceNo:发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母"c"开头，表示这笔交易被取消。
 - StockCode:产品代码。5位数，作为产品的唯一标识符
 - Description:产品名称。
 - Quantity:产品在交易中的数量
 - InvoIceDate:发票日期和时间。交易发生的日期和时间
 - UnitPrice:单价。价格单位为英镑（£）
 - CustomerID:客户编号。5位数，作为客户的唯一标识符。
 - Country:国家名称。客户所居住的国家的名称。

## **读取数据**

导入数据分析所需要的库，并通过pandas的read_csv函数，将原始数据文件"e_commerce.csv"里的数据内容，解析为DataFrame,并赋值给original_data

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## **评估数据**

在这一部分，将对在上一部分建立的original_data这个DataFrame所包含的数据进行评估  

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合"每列是一个变量，每行是一个观察值，每个单元格是一个值"这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

## **评估数据整齐度**

In [4]:
original_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
37170,539453,21876,POTTERING MUG,1,12/17/2010 17:08,3.36,NaN,United Kingdom
44013,540154,84509E,SET OF 4 CAROUSEL PLACEMATS,2,1/5/2011 11:30,3.36,NaN,United Kingdom
134717,547864,22288,HANGING METAL RABBIT DECORATION,3,3/27/2011 13:10,1.25,15716.0,United Kingdom
311738,564316,22382,LUNCH BAG SPACEBOY DESIGN,10,8/24/2011 13:11,1.65,13985.0,United Kingdom
95168,544448,22139,RETROSPOT TEA SET CERAMIC 11 PC,4,2/20/2011 12:01,4.95,14878.0,United Kingdom
489004,577869,21908,CHOCOLATE THIS WAY METAL SIGN,5,11/22/2011 11:34,2.10,13268.0,United Kingdom
3448,536630,21071,VINTAGE BILLBOARD DRINK ME MUG,6,12/2/2010 10:56,1.06,17850.0,United Kingdom
246425,558700,22553,PLASTERS IN TIN SKULLS,1,7/1/2011 12:33,1.65,17920.0,United Kingdom
143339,548687,22607,WOODEN ROUNDERS GARDEN SET,4,4/1/2011 16:50,9.95,14195.0,United Kingdom
26613,538518,22837,HOT WATER BOTTLE BABUSHKA,1,12/12/2010 16:14,4.65,14505.0,United Kingdom


从抽样的10行数据来看，数据符合"每列是一个变量，每行是一个观察值，每个单元格是一个值"，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构问题

## **评估数据干净度**

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值，而Description、CustomerID变量存在缺失值。  
此外，InvoiceDate的数据类型应该为日期，CustomerID的数据类型应该为字符串，应该进行数据格式转换。

## **评估缺失数据**

在了解Description存在缺失值后，根据条件提取出缺失观察值

In [6]:
original_data[original_data["Description"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


有1454条交易数据缺失Description变量值  
从输出结果来看，这些缺失Description的交易数据，UnitPrice都为0，为了验证猜想，我们增加筛选条件，看是否存在Description变量缺失且UnitPrice不为0的数据

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


筛选出来结果数据量为0条，说明缺失Description值的数据，同事也不具备有效的UnitPrice值。  
Description表示产品名称，UnitPrice表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这次后续可以被删除

CustomerID变量同样存在缺失值，因此也根据条件提取出缺失观察值

In [8]:
original_data[original_data["CustomerID"].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


CustomerID	表示客户编号，不是分析畅销产品的必要变量，并且从输出结果来看，有些CustomerID	缺失的交易数据仍然有效，因此保留此变量为空的观察值

## **评估重复数据**

根据数据变量的含义来看，虽然InvoiceNo、StockCode、和CustomerID都是唯一标识符，但一次交易可能包含多件商品，因此InvoiceNo可以存在重复，不同交易可以包含同件商品，因此StockCode可以存在重复，顾客可以进行多次交易或下单多个商品，因此CustomerID也可以存在重复

## **评估不一致数据**

不一致数据可能存在于Country变量中，我们要查看是否存在多个不同值指代同一国家情况

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从Country变量值来看 "USA"、"United States"均表示美国，"United Kingdom"、"UK"、"U.K"均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值

## **评估无效或错误数据**

可以通过DataFrame的describe方法，对数值统计信息进行快速了解

In [10]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


从输出结果来看，Quantity和UnitPrice存在负数，会对后续数值分析产生影响。  
因此，我们先筛选出Quantity数值为负数的观察值，进一步评估其含义

In [11]:
original_data[original_data["Quantity" ] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


从筛选结果来看，Quantity变量为负数的观察值，InvoiceNo似乎均以C开头，表示订单取消。  
但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在Quantity变量为负数且InvoiceNo不以C开头的观察值

In [12]:
original_data[(original_data["Quantity" ] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


以上猜想被证实错误，因为还存在Quantity变量为负数且InvoiceNo不以C开头的观察值  
但根据以上输出结果，这些筛选出的观察值的UnitPrice观察值均为0，因此增加UnitPrice的条件进行验证

In [13]:
original_data[(original_data["Quantity" ] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


根据输出结果，猜想得到证实，当Quantity变量为负数的时候，观察值条件满足以下条件之一：  
 - InvoiceNo以C开头，表示订单被取消
 - UnitPrice为0，说明单价为0英镑  
这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清理步骤，将Quantity为负数的观察值删除

接下来筛选出UnitPrice为负数的观察值

In [14]:
original_data[original_data["UnitPrice" ] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


从输出结果来看，UnitPrice为负数的观察值都是坏账调整，不属于实际商品交易数据，因此也在数据清理步骤中也将其删除

## **清理数据**

根据前面评估部分得到的结论，我们需要进行的数据清理包括：  
 - 把InvoiceDate变量的数据类型转换为日期时间
 - 把CustomerID变量的数据类型转换为字符串
 - 把Description变量缺失的观察值删除
 - 把Country变量值"USA"替换为"United States"
 - 把Country变量值"UK"、 "U.K"替换为"United Kingdom"
 - 把Quantity变量值为负数的观察值删除
 - 把UnitPrice变量值为负数的观察值删除

为了区分开经过清理的数据和原始数据，我们创建新的变量cleaned_data，让它为original_data复制出的副本，我们之后的清理步骤都将被运用在cleaned_data

In [15]:
cleaned_data = original_data.copy()
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


把InvoiceDate变量的数据类型转换为日期时间

In [16]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把CustomerID变量的数据类型转换为字符串

In [17]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID变量值结尾的".0"删除

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把Description变量缺失的观察值删除，并查看删除后该列空缺值个数和：

In [23]:
cleaned_data.dropna(subset=["Description"], inplace=True)

In [24]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

把Country变量值"USA"替换为"United States",并检查替换后"USA"变量值个数：

In [26]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "Untied States"})

In [27]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把Country变量值"UK"、 "U.K."替换为"United Kingdom"，并检查替换后"UK"和"U.K"变量值个数:

In [29]:
cleaned_data["Country"] = cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [31]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.k."]))

0
0


把Quantity变量值为负数的观察值删除,并检查替换后Quantity变量值为负数的个数：

In [32]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [34]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

把UnitPrice变量值为负数的观察值删除,并检查替换后UnitPrice变量值为负数的个数：

In [35]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [36]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## **保存清理后的数据**

完成数据清理后，把干净整齐的数据保存到新的文件夹里，文件名为e_commerce_cleaned.csv

In [37]:
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


In [40]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index=False)